feat(foundry): publish L2 state in `@celo/devchain-anvil`

arthurgousset opened the pull request:
### Description

1. adds script (`create_and_migrate_anvil_l2_devchain.sh`) to create, migrate, and dump L1 and L2 devchain state to distinct JSON files (`devchain.json` and `l2-devchain.json` respectively).
2. updates NPM publishing script to release L1 and L2 devchain state to [`@celo/devchain-anvil`](https://www.npmjs.com/package/@celo/devchain-anvil).

### Other changes

1. refactors Foundry scripts to use constants defined in new `constants.sh` file
1. adds and updates various yarn commands:
    ```sh
    $ yarn anvil-devchain:start-L1
    $ yarn anvil-devchain:start-L2
    $ yarn anvil-devchain:e2e-tests
    $ yarn anvil-devchain:integration-tests
    $ yarn anvil-devchain:status
    ```
1. moves `run_e2e_tests_in_anvil.sh` file to the `scripts/foundry/` directory with other Foundry-related scripts.
1. groups contract imports in `Migrations.s.sol` by Solidity version for better readability
1. small typo fixes

### Tested

<details>
<summary>Tested L2 migration script ✅ </summary>

```sh
$ yarn anvil-devchain:start-L2
# ...
Activating L2 on the devchain...
anvil_setCode
null
✨  Done in 10.90s.
```

(in a separate terminal) Check that `proxyAdminAddress` has bytecode ✅ 

```sh
$ cast rpc eth_getCode \
0x4200000000000000000000000000000000000018 \
latest \
--rpc-url=http://127.0.0.1:8546

"0x608060405234801561001057600080fd5b506...
```

(in a separate terminal) Check that `isL2()` is `true` on `CeloDistributionSchedule.sol`. ✅
For context, we can call `isL2()` on `CeloDistributionSchedule.sol` and other core contracts because they inherit from `IsL2Check.sol` 

```sh
# Get address of CeloDistributionSchedule
$ cast call \
0x000000000000000000000000000000000000ce10 \
"getAddressForStringOrDie(string calldata identifier)(address)" \
"CeloDistributionSchedule" \
--rpc-url=http://127.0.0.1:8546

0xA16cF67AFa80BB9Ce7a325597F80057c6B290fD4

# Call `isL2()` on `CeloDistributionSchedule.sol`
cast call \
0xA16cF67AFa80BB9Ce7a325597F80057c6B290fD4 \
"isL2()(bool)" \
--rpc-url=http://127.0.0.1:8546
true
```
</details>

<details>
<summary>Tested each new helper command locally from the `packages/protocol` directory ✅ </summary>

```sh
$ yarn anvil-devchain:e2e-tests

Generating and running devchain before running e2e tests...
# ...
Ran 1 test suite in 172.54ms (10.86ms CPU time): 1 tests passed, 0 failed, 0 skipped (1 total tests)
Stopping devchain...
Connection to localhost port 8546 [tcp/*] succeeded!
Killed Anvil
✨  Done in 391.47s.
```

```sh
$ yarn anvil-devchain:integration-tests

Generating and running devchain before running integration tests...
# ...
Ran 1 test suite in 375.34ms (343.75ms CPU time): 2 tests passed, 0 failed, 0 skipped (2 total tests)
Stopping devchain...
Connection to localhost port 8546 [tcp/*] succeeded!
Killed Anvil
✨  Done in 397.19s.
```

```sh
$ yarn anvil-devchain:check-is-running
# ...
Connection to localhost port 8546 [tcp/*] succeeded!
✨  Done in 0.05s.
```


</details>

### Related issues

- Fixes #11103

### Backwards compatibility

Yes, existing L1 devchain is unchanged, only adding a new L2 devchain state JSON.

### Documentation

_The set of community facing docs that have been added/modified because of this change_

arthurgousset committed:
refactor(scripts/foundry): moves `run_e2e_tests...sh` file

Small house-keeping change to keep Foundry-related bash scripts organised in a single directory. This script doesn't seem to be called from anywhere, and doesn't have a yarn command in `package.json` so it was fine to move it without any changes in other files.

arthurgousset committed:
feat(package.json): adds yarn commands for e2e and integration tests

Two helper commands to run two Foundry-related bash scripts during local development.

arthurgousset committed:
refactor(migrations_sol): groups imports for better readability

arthurgousset committed:
feat(scripts/foundry): adds MVP L2 migration script

Successfully deploys bytecode to `proxyAdminAddress`.
Does not yet dump state correctly, or call functions to active L2 correctly.

arthurgousset committed:
feat(package.json): adds two helper yarn commands

1. `anvil-devchain:start-L2`: which creates and migrates an L2 devchain
2. `anvil-devchain:check-is-running`: which can be used to check that an anvil instance is running locally and serving at localhost:8546

arthurgousset committed:
chore(migrations_sol): fix typo in function name

arthurgousset committed:
chore(migrations_sol): nit simplifies code comment for better readability

arthurgousset committed:
feat(scripts/foundry): activates CeloDistributionSchedule

Now successfully activates CeloDistributionSchedule.
Does not yet dump state correctly

arthurgousset committed:
refactor(scripts/foundry): moves variables to `constants.sh`

Simplifies maintenance and readability of scripts.

arthurgousset committed:
refactor(scripts/foundry): moves more variables to `constants.sh`

Simplifies maintenance and readability of scripts.

arthurgousset committed:
refactor(constants.sh): read bytecode from Foundry instead of Truffle

arthurgousset committed:
chore(scripts/foundry): adds TODO comment

At this point, the migration works as intended and the L2 is activated.
But, I haven't yet saved the devchain state file.
I'll need to refactor the way the L1 and L2 move the state file around.

arthurgousset commented on packages/protocol/scripts/foundry/load_and_migrate_l2_anvil_devchain.sh:
```diff
@@ -0,0 +1,76 @@
+--rpc-url http://127.0.0.1:$ANVIL_PORT
+
+# Save and rename L2 devchain state
+# mv $ANVIL_FOLDER/state.json $TMP_FOLDER/l2-devchain.json
```
As of 7315121, the migration works as intended and the L2 is activated.

```sh
cast call \
0xA16cF67AFa80BB9Ce7a325597F80057c6B290fD4 \
"isL2()(bool)" \
--rpc-url=http://127.0.0.1:8546
true
```

```sh
cast call \
0xA16cF67AFa80BB9Ce7a325597F80057c6B290fD4 \
"getCommunityRewardFraction()(uint256)" \
--rpc-url=http://127.0.0.1:8546
100000000000000000000 [1e20]
```

```sh
cast rpc eth_getCode \
0x4200000000000000000000000000000000000018 \
latest \
--rpc-url=http://127.0.0.1:8546
"0x60806040523480156100....
```

But, I haven't yet saved the devchain state file. 
I'll need to refactor the way the L1 and L2 move the state file around, because the current way doesn't work for two migrations.

arthurgousset replied:
As of e4671ebfbfbf04304e5c1a3c2731e4d8d7af2ddf, successfully dumping L2 state.

arthurgousset committed:
Merge branch 'release/core-contracts/12' into arthurgousset/feat/publish-L2-devchain

arthurgousset committed:
refactor(scripts/foundry): rename to `create_and_migrate_anvil_l2_devchain.sh`

To keep it consistent with the L1 script

arthurgousset committed:
style(migrations_sol): linting

arthurgousset committed:
chore(scripts/foundry): remove debugging statements

arthurgousset committed:
refactor(constants): adds L1 and L2 state file names

arthurgousset committed:
feat(scripts/foundry): successfully dumps L2 state

arthurgousset committed:
refactor(constants): uses `$ANVIL_RPC_URL` instead of hardcoded URL

arthurgousset committed:
refactor(constants): moves libraries to constants

arthurgousset committed:
fix(constants): adds `$PWD` to prevent errors on CI

On CI it seems like `jq` doesn't find the artifacts without an explicit path

```sh
./scripts/foundry/create_and_migrate_anvil_devchain.sh
  shell: /usr/bin/bash -e {0}
  env:
    FOUNDRY_CACHE_KEY: 2
    SUPPORTED_FOUNDRY_VERSION: nightly-f625d0fa7c51e65b4bf1e8f7931cd1c6e2e285e9
    ANVIL_PORT: 8546
jq: error: Could not open file build/contracts/Registry.json: No such file or directory
jq: error: Could not open file build/contracts/Proxy.json: No such file or directory
```

Source: https://github.com/celo-org/celo-monorepo/actions/runs/9808280627/job/27083755032#step:18:9

arthurgousset committed:
chore(workflows): debugging `jq` error on CI

Can't find build artifacts on CI, so trying to work out what might be wrong.

arthurgousset committed:
fix(constants): moves bytecode operation back into scripts

Since the constants are read before any foundry compilation occured, the constants can't read from the Foundry artifacts. Locally, I missed this because I had existing artifacts from previous compilations.

arthurgousset commented on packages/protocol/scripts/foundry/constants.sh:
```diff
@@ -0,0 +1,61 @@
+export CARBON_OFFSETTING_PARTNER="0x22579CA45eE22E2E16dDF72D955D6cf4c767B0eF"
+export CARBON_OFFSETTING_FRACTION="10000000000000000000" # 0.001 in fixidity format
+export REGISTRY_STORAGE_LOCATION="0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103" # Position is bytes32(uint256(keccak256("eip1967.proxy.admin")) - 1);
+export CELO_DISTRIBUTION_SCHEDULE_INITIAL_BALANCE="10000" # Arbitrary balance choosen at random. During the real L2 genesis, the VM will calculate and set an appropriate balance.
```
TODO: Check what balance I should manually set for the `CeloDistributionSchedule` during activation.

arthurgousset replied:
Calculated as something like = `GoldToken.CELO_SUPPLY_CAP - GoldToken.totalSupply()`

Options:
1. `totalSupply()` on devchain might be low, so might be worth fetching `totalSupply()` on Alfajores or Mainnet.
2. Set total supply to an arbitrary number:
    - 600M was allocated at genesis (that's at least in `totalSupply()` so anything above 600M)
    - 200M is supposed to be allocated in the 15 years since genesis (so 800M is ceilling)
    - Anything between 600M and 800M, like 700M.

arthurgousset replied:
`GoldToken.totalSupply()` on the L1 devchain is too low (`0`): 

Start L1 devchain (from NPM artifact):

```sh
anvil \
--port 8546 \
--state ~/Downloads/devchain-11107-2024-07-08/devchain.json
```

In a different terminal:

```sh
# Get address of GoldToken
$ cast call \
0x000000000000000000000000000000000000ce10 \
"getAddressForStringOrDie(string calldata identifier)(address)" \
"GoldToken" \
--rpc-url=http://127.0.0.1:8546
0x2365F470a061B8333BAdE225eb7B34E2f9A3B4B2

# Call `totalSupply()` on `GoldToken.sol`
$ cast call \
0x2365F470a061B8333BAdE225eb7B34E2f9A3B4B2 \
"totalSupply()(uint256)" \
--rpc-url=http://127.0.0.1:8546
0
```

arthurgousset replied:
`GoldToken.totalSupply()` on the L1 Mainnet is 695,313,643 CELO:

```sh
# Get address of GoldToken
$ cast call \
0x000000000000000000000000000000000000ce10 \
"getAddressForStringOrDie(string calldata identifier)(address)" \
"GoldToken" \
--rpc-url https://forno.celo.org
0x471EcE3750Da237f93B8E339c536989b8978a438

# Call `totalSupply()` on `GoldToken.sol`
$ cast call \
0x471EcE3750Da237f93B8E339c536989b8978a438 \
"totalSupply()(uint256)" \
--rpc-url https://forno.celo.org
695313643195035937058427065 [6.953e26]

# Convert units from wei to ether
$ cast to-unit 695313643195035937058427065 ether
695313643.195035937058427065
```

👉 **I'll manually set the amount at 700M CELO.**

For context, `GoldToken.totalSupply()` on the L1 Alfajores is not representative (109,959,073 CELO):

```sh
$ cast call \
0x000000000000000000000000000000000000ce10 \
"getAddressForStringOrDie(string calldata identifier)(address)" \
"GoldToken" \
--rpc-url https://alfajores-forno.celo-testnet.org
0xF194afDf50B03e69Bd7D057c1Aa9e10c9954E4C9

$ cast call \
0xF194afDf50B03e69Bd7D057c1Aa9e10c9954E4C9 \
"totalSupply()(uint256)" \
--rpc-url https://alfajores-forno.celo-testnet.org
109959073349267550845576790 [1.099e26]

$ cast to-unit 109959073349267550845576790 ether
109959073.349267550845576790
```

arthurgousset replied:
That's done in https://github.com/celo-org/celo-monorepo/pull/11107/commits/06e7fe747dcde8f23f05521d7cb5ae18fdc6e35b

arthurgousset replied:
I mistakenly set the initial balance of the CeloDistributionSchedule to `GoldToken.totalSupply()`, instead of `GoldToken.CELO_SUPPLY_CAP - GoldToken.totalSupply()`. This is updated and fixed here: 
- https://github.com/celo-org/celo-monorepo/pull/11140

arthurgousset committed:
chore(workflows): debugging `jq` CI error

arthurgousset committed:
fix(scripts/foundry): uses Foundry `out/` directory

Mistakenly read artifacts from Truffle's `build/` directory. Took a while to remember that Foundry write artifacts to `out/`.

Source: https://book.getfoundry.sh/reference/forge/forge-build#build-modes

arthurgousset committed:
chore(scripts/foundry): removes debugging statement

arthurgousset committed:
feat(package.json): updates yarn command names

arthurgousset committed:
docs(CONTRIBUTING): updates yarn command docs

arthurgousset committed:
chore(workflows): add back NPM install step

Not confident enough to remove it in this PR.

arthurgousset commented on .github/workflows/protocol-devchain-anvil.yml:
Note-to-self: Is the NPM install strictly necessary? 

https://github.com/celo-org/celo-monorepo/blob/c505eba576f3b4f7919cbe9af48e2610160ad115/.github/workflows/protocol-devchain-anvil.yml#L124-L126

It takes 6min to run on CI.

arthurgousset replied:
Yes, looks like it's needed by [`determine-release-version.ts`](https://github.com/celo-org/celo-monorepo/blob/1e4cc5615c51a1bed90ae045ca74a893b5b2ec0e/packages/protocol/scripts/determine-release-version.ts) and [`utils.ts`](https://github.com/celo-org/celo-monorepo/blob/1e4cc5615c51a1bed90ae045ca74a893b5b2ec0e/packages/protocol/scripts/utils.ts). 

Although we can probably do better with the amount of dependencies we install, since we don't need any of the Ganache, Truffle, and other release dependencies.

arthurgousset committed:
chore(workflows): removes code comment

arthurgousset committed:
feat(workflows): publishes L1 and L2 state

Also updates integration test to explicitly run against L1 devchain, and not L2 devchain.

arthurgousset committed:
fix(scripts/foundry): remove unnecessary `stop_anvil` call

CI doesn't like `lsof` command in `stop_anvil.sh`

```sh
/runner/_work/celo-monorepo/celo-monorepo/packages/protocol/scripts/foundry/stop_anvil.sh: line 10: lsof: command not found
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
```

Source: https://github.com/celo-org/celo-monorepo/actions/runs/9811181588/job/27092901761?pr=11107#step:19:7073

arthurgousset committed:
refactor(workflows): replace CI env with variable from `constants.sh`

Simplifies code and makes it easier to update ports in a single place.

arthurgousset committed:
fix(workflows): adds back `stop_anvil.sh` and installs `lsof` explicitly

CI fails if I don't explicitly kill an existing anvil server, so I decided to install `lsof` which is used by `stop_anvil.sh` to kill processes.

```sh
error binding to 127.0.0.1:8546: error creating server listener: Address already in use (os error 98)
```

Source: https://github.com/celo-org/celo-monorepo/actions/runs/9811346558/job/27093342967?pr=11107#step:20:30

arthurgousset committed:
chore(scripts/foundry): removes code comment

arthurgousset committed:
fix(workflows): install `lsof` in the same step

Not sure if that's the bug, but the last run where the installation and use was in different steps didn't work.

arthurgousset committed:
fix(workflows): hacky fix to run L2 anvil

I'm struggling to terminate the previous anvil server and re-start a new anvil server at the same address (local host 8546).

I tried to work around cumbersome CI environment issues (`lsof` which we use to identify the process is not available on CI).

For now, as a quick fix, I'll start anvil at an arbitrary different port 8547 and run the test against that localhost 8547.

I'll need to refactor this better.

arthurgousset commented on packages/protocol/scripts/foundry/create_and_migrate_anvil_l2_devchain.sh:
```diff
@@ -0,0 +1,78 @@
+cp $TMP_FOLDER/$L1_DEVCHAIN_FILE_NAME $TMP_FOLDER/$L2_DEVCHAIN_FILE_NAME
+
+# Start L2 devchain from L1 state and apply L2 migrations
+echo "Anvil L2 state will be saved to $TMP_FOLDER/$L2_DEVCHAIN_FILE_NAME"
```
CI has a bug here:

```
Killed Anvil
Anvil L2 state will be saved to /runner/_work/celo-monorepo/celo-monorepo/packages/protocol/.tmp/l2-devchain.json
Activating L2 on the devchain...
Error: 
error sending request for url (http://127.0.0.1:8546/): error trying to connect: tcp connect error: Connection refused (os error 111)

Context:
- Error #0: error trying to connect: tcp connect error: Connection refused (os error 111)
- Error #1: tcp connect error: Connection refused (os error 111)
- Error #2: Connection refused (os error 111)
```

Source: [CI](https://github.com/celo-org/celo-monorepo/actions/runs/9837180487/job/27154422476?pr=11107#step:19:7130)

arthurgousset replied:
The bug seems to be resolved as of this [CI run](https://github.com/celo-org/celo-monorepo/actions/runs/9838934169/job/27159898549?pr=11107) in this commit cd6ca0544b8d70a4c18f7393d3e3965829485260.

arthurgousset commented on .github/workflows/protocol-devchain-anvil.yml:
```diff
@@ -124,17 +123,35 @@ jobs:
           forge test -vvv \
           --match-path "test-sol/integration/*" \
-          --fork-url http://127.0.0.1:${{ env.ANVIL_PORT }}
+          --fork-url http://127.0.0.1:8547
```
There is a bug with this workflow on CI:

```
2024-07-08T09:46:34.824960Z ERROR foundry_evm_core::fork::init: It looks like you're trying to fork from an older block with a non-archive node which is not supported. Please try to change your RPC url to an archive node if the issue persists.
Error: 
Could not instantiate forked environment with fork url: http://127.0.0.1:8547/
Context:
- Failed to get block for block number: 256
latest block number: 256
```

Source: [CI](https://github.com/celo-org/celo-monorepo/actions/runs/9837180487/job/27154422476?pr=11107#step:20:599)

arthurgousset replied:
This seems to be a known issue
- https://github.com/foundry-rs/foundry/issues/7502

The issue is marked as fixed by this PR: 
- https://github.com/foundry-rs/foundry/pull/8160

Other similar issues: 
- https://github.com/foundry-rs/foundry/issues/7952
- https://github.com/foundry-rs/foundry/issues/7284

arthurgousset replied:
I'm still getting this bug as of this [CI run](https://github.com/celo-org/celo-monorepo/actions/runs/9838934169/job/27159898549?pr=11107) in this commit https://github.com/celo-org/celo-monorepo/commit/cd6ca0544b8d70a4c18f7393d3e3965829485260.

arthurgousset replied:
As of a906048, I implemented a workaround that circumvents the bug and passes CI. 

Basically, I'm generating and running an L1 devchain from scatch, and use that to run the integration test.

This takes longer than simply starting an L1 devchain from JSON state, but prevents the Anvil bug when loading state:
- foundry-rs/foundry#7502

arthurgousset replied:
I'll mark this as resolved but will leave a note for future reference. So this can be refactored when the bug is merged and we can bug the version.

arthurgousset commented on .github/workflows/protocol-devchain-anvil.yml:
There is a problem with the akeyless step in two workflows on CI:


1. `protocol-devchain-anvil.yml`

    ```
    Error: Failed to authenticate Post "https://api.gateway.akeyless.celo-networks-dev.org/auth": dial tcp 34.82.221.83:443: connect: connection refused
    ```
    
    Source: [CI](https://github.com/celo-org/celo-monorepo/actions/runs/9837180487/job/27154422476?pr=11107#step:15:12)

2. `publish-contracts-abi-release.yml`

    ```
    Error: Failed to authenticate Post "https://api.gateway.akeyless.celo-networks-dev.org/auth": dial tcp 34.82.221.83:443: connect: connection refused
    ```
    
    Source: [CI](https://github.com/celo-org/celo-monorepo/actions/runs/9837180480/job/27154592585?pr=11107#step:5:8)

arthurgousset replied:
From [Slack](https://clabsco.slack.com/archives/C02M9JVB34Y/p1720435680585389?thread_ts=1720432726.837389&cid=C02M9JVB34Y): 
> Akeyless cluster was under maintenance due to a [vulnerability in SSH](https://cloud.google.com/anthos/clusters/docs/security-bulletins#gcp-2024-040).
> Relaunched the wf, and it’s possible to retrieve the Akeyless secrets OK

arthurgousset replied:
Bug is fixed as of this [CI run](https://github.com/celo-org/celo-monorepo/actions/runs/9837180480/job/27156192284?pr=11107)

arthurgousset committed:
refactor(scripts/foundry): simplifying `lsof` query to find PID

arthurgousset committed:
fix(workflows): try to terminate anvil server again

arthurgousset committed:
fix(workflows): hacky workaround to prevent Anvil bug when loading state

Starting L1 from scratch instead of JSON state to circumvent this Anvil bug https://github.com/foundry-rs/foundry/issues/7502.

arthurgousset commented:
(for future reference)

This PR implements a workaround (in https://github.com/celo-org/celo-monorepo/commit/a9060489576bc930ffa84904c6c45ebcc1f36f59) to prevent a known Anvil bug when loading state.

Basically, I'm generating and running an L1 devchain from scatch, and use that to run the integration test here: 

https://github.com/celo-org/celo-monorepo/blob/a9060489576bc930ffa84904c6c45ebcc1f36f59/.github/workflows/protocol-devchain-anvil.yml#L127-L146

This takes longer than simply starting an L1 devchain from JSON state, but prevents the Anvil bug when loading state.

A better approach would be to start an anvil instance using JSON state like here: 

https://github.com/celo-org/celo-monorepo/blob/28464b6be5c774173e9eaed72ff91c1b24b57370/.github/workflows/protocol-devchain-anvil.yml#L138-L154

### Anvil bug

Known issue:
- https://github.com/foundry-rs/foundry/issues/7502
- https://github.com/foundry-rs/foundry/issues/7952
- https://github.com/foundry-rs/foundry/issues/7284

The issue (https://github.com/foundry-rs/foundry/issues/7502) is marked as fixed by this PR: 
- https://github.com/foundry-rs/foundry/pull/8160

Since we're not on the latest Foundry nightly, we won't be able to benefit from this PR until we bump our version.

arthurgousset commented on .github/workflows/protocol-devchain-anvil.yml:
I tried to manually trigger a canary `@celo/devchain-anvil` release from this branch to see if the L1 and L2 state will be published correctly. 

I used the workflow trigger and manually triggered the release in this [action](https://github.com/celo-org/celo-monorepo/actions/runs/9839996342).

Unfortunately, I can't see my canary version on NPM: [npmjs.com/package/@celo/devchain-anvil](https://www.npmjs.com/package/@celo/devchain-anvil?activeTab=versions).

I'd like to successfully release a canary version, and read the L2 state from the NPM package before merging this PR.

arthurgousset replied:
The manual workflow in this [action](https://github.com/celo-org/celo-monorepo/actions/runs/9839996342) produces artifacts that look correct. 
<img width="1150" alt="image" src="https://github.com/celo-org/celo-monorepo/assets/46296830/afb03e0f-92f2-41e7-88d4-391ee794f881">

The files in the artifact are what I would expect:

<img width="236" alt="image" src="https://github.com/celo-org/celo-monorepo/assets/46296830/41243b10-72c2-4a97-9a20-d9bc801a5e6f">

I can start anvil using the L2 state:

```sh
$ anvil --state ~/Downloads/devchain-11107-2024-07-08/l2-devchain.json
# ...
Listening on 127.0.0.1:8545
```

There is bytecode at the `proxyAdminAddress` ✅ 

```sh
$ cast rpc eth_getCode \
0x4200000000000000000000000000000000000018 \
latest \
--rpc-url=http://127.0.0.1:8545

"0x608060405234801561001057600080...
```

The `CeloDistributionSchedule` is registered ✅ 

```sh
$ cast call \
0x000000000000000000000000000000000000ce10 \
"getAddressForStringOrDie(string calldata identifier)(address)" \
"CeloDistributionSchedule" \
--rpc-url=http://127.0.0.1:8545

0xA16cF67AFa80BB9Ce7a325597F80057c6B290fD4
```

`isL2` is `true` on the `CeloDistributionSchedule` ✅ 

```sh
# Call `isL2()` on `CeloDistributionSchedule.sol`
$ cast call \
0xA16cF67AFa80BB9Ce7a325597F80057c6B290fD4 \
"isL2()(bool)" \
--rpc-url=http://127.0.0.1:8545
true
```

arthurgousset replied:
Turns out the reason is that a [previous step](https://github.com/celo-org/celo-monorepo/actions/runs/9839996342/job/27163239971#step:22:1) ("[Determine release type and version (or dry run)](https://github.com/celo-org/celo-monorepo/blob/a9060489576bc930ffa84904c6c45ebcc1f36f59/.github/workflows/protocol-devchain-anvil.yml#L159)") sets the `$DRY_RUN="--dry-run"` flag.

That means the package is not published.

```
$ npm publish $RELEASE_TYPE $DRY_RUN --access public
```

Source: [CI](https://github.com/celo-org/celo-monorepo/actions/runs/9839996342/job/27163239971#step:25:3)

arthurgousset replied:
@soloseng kindly agreed to investigate why that's the case and open a separate PR for this.

soloseng replied:
I was able to publish without any changes. @celo/devchain-anvil@2.0.0-layer-two-devchain-test.0


The issue was because the tag you used [didn't match the RegEx](https://github.com/celo-org/celo-monorepo/blob/a9060489576bc930ffa84904c6c45ebcc1f36f59/packages/protocol/scripts/utils.ts#L54). No numbers allowed.

arthurgousset replied:
Great catch, thanks for investigating this 🙌 I confirm I see a test release on NPM: [2.0.0-layer-two-devchain-test.0](https://www.npmjs.com/package/@celo/devchain-anvil/v/2.0.0-layer-two-devchain-test.0)

soloseng commented on .github/workflows/protocol-devchain-anvil.yml:
```diff
@@ -124,17 +123,31 @@ jobs:
+
+          ./scripts/foundry/stop_anvil.sh
+          
+      - name: Generate L1 and L2 migrations
```
NIT:

```suggestion
      - name: Generate L2 migration
```

soloseng commented on .github/workflows/protocol-devchain-anvil.yml:
```diff
@@ -124,17 +123,31 @@ jobs:
+      # Install `lsof` dependency, because it's not readily available on CI, but is required by 
+      # `create_and_migrate_anvil_l2_devchain.sh`, because it uses `stop_anvil.sh` to kill 
+      # existing anvil servers.
+      - name: Run integration tests against L1 devchain
```
NIT:

```suggestion
      - name: Generate L1 migrations and run integration tests against L1 devchain
```

soloseng commented on .github/workflows/protocol-devchain-anvil.yml:
```diff
@@ -124,17 +123,31 @@ jobs:
+          
+      - name: Generate L1 and L2 migrations
+        if: success() || failure()
+        run: ./scripts/foundry/create_and_migrate_anvil_l2_devchain.sh
```
Might be good to create an issue to track adding integration test here. 

I briefly looked at the error that happens when running the integration test on L2, and its because `RevokeCeloAfterL2TransitionTest` assumes that the chain starts on L1 than transitions to L2 within the test. 

This causes the validator contract to fail during initialization since devchain L2 is already active.

arthurgousset replied:
Great point, that's done here: 
- https://github.com/celo-org/celo-monorepo/issues/11138

martinvol commented:
is there a script within this PR where you can import a L1 state and it'll dump the state after the L2?

martinvol commented on packages/protocol/scripts/foundry/create_and_migrate_anvil_l2_devchain.sh:
```diff
@@ -0,0 +1,78 @@
+--rpc-url $ANVIL_RPC_URL
+
+# Fetch address of Celo distribution 
+CELO_DISTRIBUTION_SCHEDULE_ADDRESS=$(
```
I'd replace this file with a forge script (like Migrations.s.sol), so we have all this code already written in solidity and we can combine it to just one migration after the mainnet transition is done.

arthurgousset replied:
Great point, I propose:
1. we merge this PR as is, that way DevTooling can get started using the L2 devchain. 
2. we open an issue to refactor this workflow into a Foundry script.

Here is an issue to track this:
- https://github.com/celo-org/celo-monorepo/issues/11139

arthurgousset committed:
Update .github/workflows/protocol-devchain-anvil.yml

Co-authored-by: soloseng <102702451+soloseng@users.noreply.github.com>

arthurgousset committed:
Update .github/workflows/protocol-devchain-anvil.yml

Co-authored-by: soloseng <102702451+soloseng@users.noreply.github.com>

arthurgousset commented:
> is there a script within this PR where you can import a L1 state and it'll dump the state after the L2?

@martinvol: Good question, not in the current state. 

There is a limitation in Foundry that prevents us from loading state, and then dumping state to a different file. For example:

```sh
$ anvil \
--port 8546 \
--state ~/Downloads/devchain-11107-2024-07-08/devchain.json \
--dump-state ~/Downloads/l2-devchain.json

error: the argument '--state <PATH>' cannot be used with '--dump-state <PATH>'
```

But, this was a design I considered during this PR. We can circumvent this restriction by copying and renaming files during runtime like we do for L1:

https://github.com/celo-org/celo-monorepo/blob/ff291e008f85418bccc825911aeed03048ece125/packages/protocol/scripts/foundry/create_and_migrate_anvil_devchain.sh#L67-L68

But, I took the view that it was overkill because it also required refactoring `start_anvil.sh` which specifies the name of the JSON state file. That's because we want two JSON files, L1 and L2. 

https://github.com/celo-org/celo-monorepo/blob/ff291e008f85418bccc825911aeed03048ece125/packages/protocol/scripts/foundry/start_anvil.sh#L23-L26

Definitely doable, but for the scope of this PR, which was to release an L2 devchain, I found the current approach to be the fastest path to an L2 devchain without significantly re-organising our scripts.

We can refactor our bash and Foundry scripts to work differently if useful.

arthurgousset committed:
feat(constants): set CeloDistributionSchedule initial balance to 700M CELO

Arbitrary amount chosen to be approximately equal to `GoldToken.totalSupply()` on the L1 Mainnet (695,313,643 CELO as of this commit). During the real L2 genesis, the VM will calculate and set an appropriate balance.

```sh
# Get address of GoldToken
$ cast call \
0x000000000000000000000000000000000000ce10 \
"getAddressForStringOrDie(string calldata identifier)(address)" \
"GoldToken" \
--rpc-url https://forno.celo.org
0x471EcE3750Da237f93B8E339c536989b8978a438

# Call `totalSupply()` on `GoldToken.sol`
$ cast call \
0x471EcE3750Da237f93B8E339c536989b8978a438 \
"totalSupply()(uint256)" \
--rpc-url https://forno.celo.org
695313643195035937058427065 [6.953e26]

# Convert units from wei to ether
$ cast to-unit 695313643195035937058427065 ether
695313643.195035937058427065
```

arthurgousset committed:
docs(README): improves README

More context, examples, background, files in the package.
